File: 2022/day15.go
package main

import (
	"bufio"
	"image"
	"log"
	"os"
	"regexp"
	"sort"
	"strconv"
)

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func manhattan(a image.Point, b image.Point) int {
	return abs(a.X-b.X) + abs(a.Y-b.Y)
}

type Range struct {
	start int
	stop  int
}

func (r *Range) contains(x int) bool {
	return r.start <= x && x < r.stop
}

func (r *Range) length() int {
	return r.stop - r.start
}

func (r *Range) union(other *Range) *Range {
	// overlapping ranges
	if r.contains(other.start) {
		if other.stop > r.stop {
			return &Range{r.start, other.stop}
		}
		return r
	}
	if other.contains(r.start) {
		if r.stop > other.stop {
			return &Range{other.start, r.stop}
		}
		return other
	}
	// adjoining ranges
	if other.start == r.stop {
		return &Range{r.start, other.stop}
	}
	if r.start == other.stop {
		return &Range{other.start, r.stop}
	}
	// disjoint
	return nil
}

func appendRange(ranges []Range, newRange *Range) []Range {
	for i, r := range ranges {
		u := r.union(newRange)
		if u == &r {
			return ranges
		}
		if u != nil {
			ranges = append(ranges[:i], ranges[i+1:]...)
			return appendRange(ranges, u)
		}
	}
	return append(ranges, *newRange)
}

func collectRanges(sensors map[image.Point]int, y int) []Range {
	var ranges []Range
	for sensor, strength := range sensors {
		strength -= abs(y - sensor.Y)
		if strength >= 0 {
			ranges = appendRange(ranges, &Range{sensor.X - strength, sensor.X + strength + 1})
		}
	}
	return ranges
}

func main() {
	sensors := make(map[image.Point]int)
	beacons := make(map[image.Point]bool)

	file, err := os.Open("day15_input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	pattern := regexp.MustCompile("Sensor at x=(-?\\d+), y=(-?\\d+): closest beacon is at x=(-?\\d+), y=(-?\\d+)")
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		match := pattern.FindStringSubmatch(scanner.Text())
		if len(match) != 5 {
			log.Fatal("Expected 5 matches, got ", len(match))
		}
		sx, err := strconv.Atoi(match[1])
		if err != nil {
			log.Fatal(err)
		}
		sy, err := strconv.Atoi(match[2])
		if err != nil {
			log.Fatal(err)
		}
		bx, err := strconv.Atoi(match[3])
		if err != nil {
			log.Fatal(err)
		}
		by, err := strconv.Atoi(match[4])
		if err != nil {
			log.Fatal(err)
		}
		sensor := image.Pt(sx, sy)
		beacon := image.Pt(bx, by)
		distance := manhattan(sensor, beacon)
		sensors[sensor] = distance
		beacons[beacon] = true
	}

	part1 := 0
	for _, r := range collectRanges(sensors, 2000000) {
		part1 += r.length()
	}
	for b := range beacons {
		if b.Y == 2000000 {
			part1 -= 1
		}
	}
	println(part1)

	xs := Range{0, 4000000 + 1}
	for y := 0; y <= 4000000; y++ {
		ranges := collectRanges(sensors, y)
		sort.Slice(ranges, func(i, j int) bool {
			return ranges[i].start < ranges[j].start
		})
		for _, r := range ranges {
			if r.union(&xs) != &r {
				x := r.stop
				println(4000000*x + y)
				return
			}
		}
	}
}

File: 2022/day15.py
import doctest
import re


def print_grid(grid):
    minx = min(int(c.real) for c in grid.keys())
    maxx = max(int(c.real) for c in grid.keys())
    miny = min(int(c.imag) for c in grid.keys())
    maxy = max(int(c.imag) for c in grid.keys())

    for y in range(miny, maxy + 1):
        for x in range(minx, maxx + 1):
            print(grid.get(x + y * 1j, '.'), end='')
        print('')


def manhattan(a, b):
    return int(abs(a.real - b.real) + abs(a.imag - b.imag))


def diamond(start, distance):
    for y in range(-distance, distance + 1):
        for x in range(-(distance - abs(y)), (distance - abs(y)) + 1):
            yield start + x + y * 1j


def main():
    sensors = dict()
    beacons = dict()

    p = re.compile(r'Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)')
    with open('day15_input.txt') as f:
        for line in f:
            match = p.match(line)
            if match:
                [sx, sy, bx, by] = [int(v) for v in match.groups()]
                sensor = sx + sy * 1j
                beacon = bx + by * 1j
                distance = manhattan(sensor, beacon)
                sensors[sensor] = distance
                beacons[beacon] = True

    minx = int(min(k.real - v for k, v in sensors.items()))
    maxx = int(max(k.real + v for k, v in sensors.items()))

    def in_range_of_sensor(pos):
        return any(manhattan(pos, k) <= v for k, v in sensors.items())

    def not_a_beacon(pos):
        return pos not in beacons and in_range_of_sensor(pos)

    y = 2000000
    print(sum(1 for x in range(minx, maxx + 1) if not_a_beacon(x + y * 1j)))


if __name__ == "__main__":
    doctest.testmod()
    main()
